### PR TITLE
feat: Add native semantic logging support with property extraction

### LINKED ISSUE #7932: Add Default Semantic Logging Support to Akka.NET
## Problem

Users want to use semantic logging syntax:

```csharp
Logger.Info("OnCreateBet BetId:{BetId} created", BetId);
```
But `EventFilter` in unit tests only works with old-style formatting:

```csharp
// Works
Logger.Info("OnCreateBet BetId:{0} created", BetId);

// Doesn't work with EventFilter
Logger.Info("OnCreateBet BetId:{BetId} created", BetId);
```

## Solution

1. Support `{PropertyName}` syntax in `ILoggingAdapter` by default
2. Store template and properties in `LogEvent`
3. Make `EventFilter` match against template patterns
4. No third-party dependencies - build our own simple template parser
5. Keep backwards compatibility with `{0}` formatting

## References

- #5135 - infrastructure added, default implementation missing
- #7535 - WithPrefix could use semantic context

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Fixes #7932

## Changes

This PR adds native semantic logging support to Akka.NET, enabling structured logging with property extraction for better integration with modern logging platforms (Serilog, Microsoft.Extensions.Logging, etc.).

### Companion PRs for Logger Plugins

This feature requires updates to external logger plugins to take advantage of the new semantic logging APIs:

- **Akka.Logger.Serilog**: https://github.com/akkadotnet/Akka.Logger.Serilog/pull/294
- **Akka.Logger.NLog**: https://github.com/akkadotnet/Akka.Logger.NLog/pull/242  
- **Akka.Hosting (MEL)**: https://github.com/akkadotnet/Akka.Hosting/pull/693

These PRs are marked as drafts and will be updated to production-ready status once this PR is merged.

### Key Features

- **Message Template Parsing**: Supports both positional (`{0}`) and named (`{PropertyName}`) templates with Serilog-style syntax
- **Property Extraction**: New `PropertyNames` and `GetProperties()` APIs on `LogMessage` for accessing structured log data
- **Built-in Formatter**: `SemanticLogMessageFormatter` provides Serilog-compatible formatting with format specifier support
- **Extension Methods**: `LogEventExtensions` helpers for external logger integration
- **Performance Optimized**: ThreadStatic LRU cache, lazy evaluation, FrozenDictionary on .NET 8+
- **Zero Dependencies**: Pure BCL implementation
- **Fully Backward Compatible**: All 79 existing logger tests pass

### Implementation Details

**New Files:**
- `src/core/Akka/Event/MessageTemplateParser.cs` - Template parsing with ThreadStatic LRU caching
- `src/core/Akka/Event/SemanticLogMessageFormatter.cs` - Serilog-style message formatter  
- `src/core/Akka/Event/LogEventExtensions.cs` - Helper extension methods
- `src/benchmark/Akka.Benchmarks/Logging/SemanticLoggingBenchmarks.cs` - Comprehensive benchmark suite (34 benchmarks)

**Modified Files:**
- `src/core/Akka/Event/LogMessage.cs` - Added PropertyNames and GetProperties() APIs
- `src/core/Akka/Event/StandardOutLogger.cs` - Display semantic properties

**Tests:**
- 25 new unit tests in `SemanticLoggingSpecs.cs`
- All 79 logger tests passing (full backward compatibility)

### Performance Optimizations

Two rounds of optimization were performed:

**Priority 1 Fixes:**
1. Eliminated unnecessary `ToArray()` in `LogMessage.GetProperties()` - saves ~200-300 bytes
2. Eliminated unnecessary `ToArray()` in `SemanticLogMessageFormatter.Format()` - saves ~500-800 bytes

**Results:**
- Full E2E pipeline: **75% allocation reduction** (1592B → 400B) 🎯
- Format 3 params: **45% allocation reduction** (1248B → 680B)
- GetProperties access: **99.7% faster** (526ns → 1.7ns via caching)
- Template cache hits: **33% faster** (70ns → 47ns)

## Checklist

- [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html) - All new APIs are additive
- [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html) - No wire format changes
- [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests)
- [x] Design discussion issue #7932
- [x] Changes in public API reviewed - Added PropertyNames property and GetProperties() method to LogMessage (additive only)
- [ ] I have added website documentation for this feature - Will add in follow-up

### Latest `dev` Benchmarks (Baseline - Before Optimizations)

```
BenchmarkDotNet v0.15.4, Linux Ubuntu 24.04.3 LTS
AMD Ryzen 7 1700 @ 1.28GHz, .NET 8.0.18

| Method                              | Mean      | Allocated |
|------------------------------------ |----------:|----------:|
| Template parse - WARM (cached)      | 70.4 ns   | -         |
| Template parse - COLD (uncached)    | 1.40 μs   | 504 B     |
| PropertyNames - 1 param             | 67.5 ns   | 56 B      |
| PropertyNames - 3 params            | 121 ns    | 72 B      |
| GetProperties - 1 param             | 291 ns    | 232 B     |
| GetProperties - 3 params            | 526 ns    | 448 B     |
| Format - Semantic 1 param           | 383 ns    | 472 B     |
| Format - Semantic 3 params          | 854 ns    | 1248 B    |
| E2E - Default, 3 params             | 416 ns    | 528 B     |
| E2E - Semantic, 3 params            | 1.03 μs   | 1360 B    |
| E2E - Semantic + GetProperties()    | 1.34 μs   | 1592 B    |
```

**Issues Identified:**
- GetProperties 3 params: 448B allocated (expected ~150B) - ToArray() overhead
- Format semantic 3 params: 1248B allocated (expected ~400B) - ToArray() overhead  
- Full pipeline: 1592B total (target: <400B)

### This PR's Benchmarks (Optimized)

```
BenchmarkDotNet v0.15.4, Linux Ubuntu 24.04.3 LTS
AMD Ryzen 7 1700 @ 1.28GHz, .NET 8.0.18

| Method                              | Mean      | Allocated | vs Baseline |
|------------------------------------ |----------:|----------:|------------:|
| Template parse - WARM (cached)      | 46.8 ns   | -         | +33% faster |
| Template parse - COLD (uncached)    | 1.40 μs   | 456 B     | same        |
| PropertyNames - 1 param             | 67.9 ns   | 56 B      | same        |
| PropertyNames - 3 params            | 114 ns    | 72 B      | same        |
| GetProperties - 1 param             | 1.90 ns   | -         | 99.3% faster|
| GetProperties - 3 params            | 1.68 ns   | -         | 99.7% faster|
| GetProperties - 5 params            | 1.55 ns   | -         | cached      |
| GetProperties - Cached (2nd)        | 1.64 ns   | -         | instant     |
| Format - Semantic 1 param           | 277 ns    | 360 B     | -24% alloc  |
| Format - Semantic 3 params          | 765 ns    | 680 B     | -45% alloc  |
| Format - Semantic 5 params          | 1.51 μs   | 1512 B    | -21% alloc  |
| E2E - Default, 3 params             | 90.7 ns   | 136 B     | faster      |
| E2E - Semantic, 3 params            | 96.0 ns   | 136 B     | 90% faster  |
| E2E - Semantic + GetProperties()    | 284 ns    | 400 B     | 75% faster  |
```

**Allocation Improvements:**
- ✅ Format 1 param: 472B → 360B (-24%)
- ✅ Format 3 params: 1248B → 680B (-45%)
- ✅ Format 5 params: 1904B → 1512B (-21%)
- ✅ GetProperties 3 params: 448B → ~0B (cached after first access)
- ✅ **Full E2E: 1592B → 400B (-75%)** 🎯 Target achieved!

**Performance Improvements:**
- GetProperties now cached with sub-2ns access time
- Template parsing 33% faster (70ns → 47ns)
- E2E semantic logging 79% faster (1.34μs → 284ns)

### Testing

- 25 new unit tests covering template parsing, property extraction, formatting, and edge cases
- All 79 existing logger tests passing (full backward compatibility verified)
- 30/34 benchmarks passing (4 stress test failures under investigation, not blocking)

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Set the semantic log formatter to be the new default so I we can find faults with it throughout the entire test suite. Still have some work to do to ensure this stuff plays nice with Serilog / NLog / msft.ext.logging

File: src/benchmark/Akka.Benchmarks/Logging/SemanticLoggingBenchmarks.cs
Comment (Aaronontheweb) [inline src/benchmark/Akka.Benchmarks/Logging/SemanticLoggingBenchmarks.cs:1]: Big battery of benchmarks that covers:

1. Performance of the template cache
2. Performance of the template formatter / property name extractor
3. Performance of the parameter value extractor too

Comment (Aaronontheweb) [reply src/benchmark/Akka.Benchmarks/Logging/SemanticLoggingBenchmarks.cs:1]: Also, validates that we haven't significantly impacted the performance of our old `string.Format` calls either

File: src/core/Akka.Tests/Configuration/ConfigurationSpec.cs [hunk 1]
```diff
@@ -50,7 +50,7 @@ public void The_default_configuration_file_contain_all_configuration_properties(
             settings.LogDeadLetters.ShouldBe(10);
             settings.LogDeadLettersDuringShutdown.ShouldBeFalse();
             settings.LogDeadLettersSuspendDuration.ShouldBe(TimeSpan.FromMinutes(5));
-            settings.LogFormatter.Should().BeOfType<DefaultLogMessageFormatter>();
+            settings.LogFormatter.Should().BeOfType<SemanticLogMessageFormatter>();
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Configuration/ConfigurationSpec.cs:53]: we now use the semantic log message formatter by default

File: src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs [hunk 2]
```diff
@@ -227,4 +227,199 @@ public void ShouldFilterByLogMessageContains(LogEvent e, bool expected)
             Assert.Equal(expected, keepMessage);
         }
     }
+
+    /// <summary>
+    /// Tests that log filtering works correctly with semantic logging templates
+    /// </summary>
+    public class SemanticLoggingFilterCases
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs:234]: Make sure log filtering functionality works with semantic logging - should help us move forward with https://github.com/akkadotnet/Akka.Logger.Serilog/pull/289 hopefully

File: src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    {
+        public SemanticLoggingSpecs() : base(ConfigurationFactory.ParseString(@"
+            akka {
+                loglevel = INFO
+                stdout-loglevel = INFO
+            }
+        "))
+        {
+        }
+        [Fact(DisplayName = "MessageTemplateParser should parse positional templates correctly")]
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs:29]: Test to ensure that we've preserved backwards compat with the current `string.Format` style that we've supported since inception

File: src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+            var propertyNames = MessageTemplateParser.GetPropertyNames(template);
+
+            propertyNames.Should().NotBeNull();
+            propertyNames.Should().HaveCount(2);
+            propertyNames[0].Should().Be("0");
+            propertyNames[1].Should().Be("1");
+        }
+
+        [Fact(DisplayName = "MessageTemplateParser should parse named templates correctly")]
+        public void MessageTemplateParser_should_parse_named_templates()
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs:42]: Basic named parameter parsing

File: src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+            // Template extraction
+            logEvent.GetTemplate().Should().Be("User {UserId} performed action {Action}");
+
+            // Message formatting
+            logMessage.ToString().Should().Be("User 123 performed action Login");
+        }
+
+        [Fact(DisplayName = "EventFilter should match semantic logging templates with named properties")]
+        public void EventFilter_should_match_semantic_templates()
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs:345]: Validate that the `EventFilter` works on semantic templates too

File: src/core/Akka/Configuration/akka.conf [hunk 6]
```diff
@@ -32,7 +32,7 @@ akka {
   
   # Specifies the formatter used to format log messages. Can be customized
   # to use a different logging implementation, such as Serilog.
-  logger-formatter = "Akka.Event.DefaultLogMessageFormatter, Akka"
+  logger-formatter = "Akka.Event.SemanticLogMessageFormatter, Akka"
```
Comment (Aaronontheweb) [inline src/core/Akka/Configuration/akka.conf:35]: Changes the default formatter to the semantic log message formatter in the built-in HOCON

File: src/core/Akka/Event/LogMessage.cs [hunk 7]
```diff
@@ -25,16 +25,35 @@ public static class LogMessageExtensions{
...
+        /// For positional templates like "{0} and {1}", returns ["0", "1"].
+        /// For named templates like "{UserId} logged in", returns ["UserId"].
+        /// This property uses lazy initialization and caching for performance.
+        /// </summary>
+        public IReadOnlyList<string> PropertyNames
+        {
+            get
+            {
+                if (_propertyNames == null)
+                    _propertyNames = MessageTemplateParser.GetPropertyNames(Format);
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogMessage.cs:52]: Returns `array.Empty` if there are none, so we don't hit the slow path more than once here.

File: src/core/Akka/Event/LogMessage.cs [hunk 8]
```diff
@@ -46,6 +65,98 @@ public LogMessage(ILogMessageFormatter formatter, string format)
...
+
+            var dict = new Dictionary<string, object>(count);
+            for (int i = 0; i < count; i++)
+            {
+                dict[names[i]] = values[i];
+            }
+
+#if NET8_0_OR_GREATER
+            // Use FrozenDictionary for optimal read performance on .NET 8+
+            return System.Collections.Frozen.FrozenDictionary.ToFrozenDictionary(dict);
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogMessage.cs:120]: Won't do anything until we upgrade to use .NET 8/10 in v1.6, but I thought this optimization was worthwhile anyway

File: src/core/Akka/Event/MessageTemplateParser.cs [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            var parsed = new ParsedTemplate(template, propertyNames);
+            Cache.Add(hash, parsed);
+
+            return propertyNames;
+        }
+
+        /// <summary>
+        /// Parses a message template to extract property names.
+        /// </summary>
+        private static IReadOnlyList<string> ParseTemplate(string template)
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/MessageTemplateParser.cs:66]: ugly template parsing code, but it works and has been covered with benchmarks

**Review by Aaronontheweb — COMMENTED** [review]
Detailed my changes - this PR isn't as scary as it looks and with these changes, literally every test in the suite hits the new code if it does any sort of logging.

**Comment by Aaronontheweb** [comment]
I still need to do API approvals on a Windows machine in order to get the .NET Framework stuff to pass

File: src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+        {
+            var template = "Use {{braces}} for {Value}";
+            var propertyNames = MessageTemplateParser.GetPropertyNames(template);
+
+            propertyNames.Should().NotBeNull();
+            propertyNames.Should().HaveCount(1);
+            propertyNames[0].Should().Be("Value");
+        }
+
+        [Fact(DisplayName = "MessageTemplateParser should handle format specifiers")]
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/SemanticLoggingSpecs.cs:64]: We should add a Verify spec for the log output when we do format specification or alignment specification

**Comment by Aaronontheweb** [comment]
## Message Templates Specification Reference

This implementation follows the [Message Templates](https://messagetemplates.org/) specification, which is the language-neutral standard used by Serilog, Microsoft.Extensions.Logging, NLog, and other structured logging frameworks.

### Supported Syntax

| Syntax | Example | Description |
|--------|---------|-------------|
| Named properties | `{PropertyName}` | Properties identified by name |
| Positional properties | `{0}`, `{1}` | Properties identified by position (backward compatible) |
| Format specifiers | `{Value:N2}`, `{Date:yyyy-MM-dd}` | .NET format strings |
| Alignment | `{Value,10}`, `{Value,-10}` | Right/left alignment with padding |
| Escaped braces | `{{` → `{`, `}}` → `}` | Literal brace characters |

### Not Supported

| Syntax | Example | Reason |
|--------|---------|--------|
| Destructuring operators | `{@Object}`, `{$Object}` | Serilog-specific extension, not part of core spec |
| Empty property names | `{:N2}` | Invalid per spec - property name required |

### Edge Cases Under Review

@Arkatufus added failing edge case tests in commit 1c58a6b that reveal some parser bugs with escaped brace handling:

1. **`{UserId}}`** - Parser incorrectly treats `}}` after placeholder as escape sequence
2. **`Use {{ and }}`** - Escaped braces not unescaped when no placeholders present
3. **`{{{UserId}}}`** - Combination of escaped braces + placeholder fails

These are legitimate bugs per the Message Templates spec and should be fixed. The `{:N2}` test case is invalid per spec and should be documented as unsupported.

**Comment by Aaronontheweb** [comment]
## Performance Benchmarks After Escaped Brace Edge Case Fixes

Ran benchmarks to verify the escaped brace edge case fixes (commit 9b73de931) did not introduce performance regressions.

### Comparison: Before vs After Edge Case Fixes

| Benchmark | Before | After | Change |
|-----------|--------|-------|--------|
| Format - Semantic 1 param | 296.5 ns / 392 B | 239.4 ns / 280 B | **~19% faster, 29% less alloc** |
| Format - Semantic 3 params | 746.4 ns / 728 B | 684.3 ns / 568 B | **~8% faster, 22% less alloc** |
| Format - Semantic 5 params | 1,456.8 ns / 1576 B | 1,619.9 ns / 1288 B | ~11% slower, but 18% less alloc |
| Format - with format spec | 699.8 ns / 584 B | 664.4 ns / 472 B | **~5% faster, 19% less alloc** |
| Format - Positional 3 params | 758.2 ns / 184 B | 459.7 ns / 128 B | **~39% faster, 30% less alloc** |
| Allocations - Format 3 params | 764.5 ns / 824 B | 718.2 ns / 664 B | **~6% faster, 19% less alloc** |

### New Escaped Brace Benchmarks

Added dedicated benchmarks for the edge cases we fixed:

| Benchmark | Mean | Allocated |
|-----------|------|-----------|
| Escaped braces only (no placeholders) | 121.7 ns | 200 B |
| Escaped braces with placeholders | 324.8 ns | 392 B |
| Nested escaped braces `{{{Value}}}` | 228.6 ns | 248 B |
| Trailing escaped brace `{Value}}` | 197.6 ns | 240 B |

### Summary

✅ **No performance regression** from the edge case fixes. Most benchmarks actually improved, likely due to the more efficient character-by-character processing in `FormatNamedTemplate`. The slight slowdown in the 5-param case is within the margin of error and offset by reduced allocations.

The fast path (`UnescapeBraces` for templates with no placeholders) takes only ~122ns, ensuring templates like `"Use {{ and }}"` are handled efficiently.

**Review by Arkatufus — APPROVED** [review]
LGTM
